libndt7/src/main/java/net/measurementlab/ndt7/android/ClientUpload.java

```diff
@@ -0,0 +1,237 @@
+            uri = new URI(
+                    "wss",
+                    null, // userInfo
+                    settings.getHostname(),
+                    (settings.getPort() >= 0 && settings.getPort() < 65536) ? settings.getPort() : -1,
+                    "/ndt/v7/upload",
+                    "",
+                    null
+            );
+        } catch (URISyntaxException e) {
+            Log.e(TAG, "runUpload encountered exception", e);
+            onError(e.getMessage());
+            return false;
+        }
+
+        OkHttpClient.Builder builder = new OkHttpClient.Builder();
+        if (settings.getSkipTlsCertificateVerification()) {
+            X509TrustManager x509TrustManager = new X509TrustManager() {
+                @Override
+                public void checkClientTrusted(X509Certificate[] chain,
+                                               String authType) { }
+
+                @Override
+                public void checkServerTrusted(X509Certificate[] chain,
+                                               String authType) { }
+
+                @Override
+                public X509Certificate[] getAcceptedIssuers() {
+                    return new X509Certificate[]{};
+                }
+            };
+
+            try {
+                final TrustManager[] trustAllCerts = new TrustManager[]{ x509TrustManager };
+                final SSLContext sslContext = SSLContext.getInstance("SSL");
+                sslContext.init(null, trustAllCerts, new java.security.SecureRandom());
+                final SSLSocketFactory sslSocketFactory = sslContext.getSocketFactory();
+                builder.sslSocketFactory(sslSocketFactory, x509TrustManager);
+            } catch (Exception e) {
+                Log.e(TAG, "Encountered exception", e);
+            }
+
+            builder.hostnameVerifier((hostname, session) -> true);
+        }
+
+        OkHttpClient client = builder
+                .connectTimeout(30, TimeUnit.SECONDS)
+                .readTimeout(30, TimeUnit.SECONDS)
+                .writeTimeout(40, TimeUnit.SECONDS)
+                .build();
+
+
+
+        Request request = new Request.Builder()
+                .url(uri.toString())
+                .addHeader("Sec-WebSocket-Protocol", "net.measurementlab.ndt.v7")
+                .method("GET",null)
+                .build();
+        Log.d(TAG,request.toString());
+        client.newWebSocket(request, this);
+
+        t0 = tLast = System.nanoTime();
+        client.dispatcher().executorService().shutdown();
+
+        // Basically make the code synchronous here:
+
+        /*ExecutorService svc = client.dispatcher().executorService();
+        svc.shutdown();
+        try {
+            svc.awaitTermination(30, TimeUnit.SECONDS);
+        } catch (InterruptedException e) {
+            // TODO(bassosimone): how to handle this error condition?
+            Log.e(TAG, "runDownload awaitTermination encountered exception", e);
+        }*/
```

Comment:
I see you commented out similar code also in the download path. It seems I'm failing to guess why. Can you explain?

Reply:
Hi @bassosimone ,
I did it because with this code you make a blocking call and the UI will be unresponsive untill the download/upload is completed.

Reply:
Gotcha! Then, can you please remove the commented out code? I do not believe there is much value for us keeping that around, right?

Reply:
Yes, sure!